pkg/controllers/workgenerator/controller.go

```diff
@@ -283,10 +284,28 @@ func (r *Reconciler) ensureFinalizer(ctx context.Context, resourceBinding client
 	if controllerutil.ContainsFinalizer(resourceBinding, fleetv1beta1.WorkFinalizer) {
 		return nil
 	}
-	controllerutil.AddFinalizer(resourceBinding, fleetv1beta1.WorkFinalizer)
-	if err := r.Client.Update(ctx, resourceBinding); err != nil {
-		klog.ErrorS(err, "Failed to add the work finalizer to resourceBinding", "resourceBinding", klog.KObj(resourceBinding))
-		return controller.NewUpdateIgnoreConflictError(err)
+
+	// Add retries to the update behavior as the binding object can become a point of heavy
+	// contention under heavy workload; simply requeueing when a write conflict occurs, though
+	// functionally correct, might trigger the work queue rate limiter and eventually lead to
+	// substantial delays in processing.
+	//
+	// Also note that here default backoff strategy (exponetial backoff) rather than the Kubernetes'
+	// recommended on-write-conflict backoff strategy is used, as experimentation suggests that
+	// this backoff strategy yields better performance, especially for the long-tail latencies.
+	//
+	// TO-DO (chenyu1): evaluate if a custom backoff strategy can get an even better result.
+	errAfterRetries := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
This is not the only place the work generator needs to update the object, is there any reason not to change other parts?

Reply:
From my end of the test this is the only update that being blocked heavily by the rate limiter; the other update seems to be fine using the queue's retry behavior.